<a href="https://colab.research.google.com/github/ErnestoS215/GLAB-386.7.1---Exploring-Airbnb-Market-Trends-using-AI-assistance/blob/main/GLAB_386_7_1_Exploring_Airbnb_Market_Trends_using_AI_assistance.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Guided Lab 386.7.1 - Exploring Airbnb Market Trends using AI Assistance**

#**Lab Overview**
In this lab, we will explore and analyze a simplified Airbnb dataset to understand market trends, pricing patterns, room-type distribution, and review activity.

We will work with multiple datasets in different file formats and combine them into a single consolidated DataFrame using a common key. After merging the data, we will clean and standardize columns to ensure consistency and accuracy.

We will use AI coding assistance tools (such as Google Colab AI, ChatGPT, or GitHub Copilot) to help generate code for routine tasks like data cleaning, summary statistics, and visualization. This will allow us to focus more on interpreting the results, identifying meaningful insights, and communicating our findings clearly.

##**AI tools:**
For this lab you can leverage AI tools such asGoogle Colab AI button, ChatGPT, GitHub Copilot, or similar assistants.

#**Learning objectives**

By the end of this lab, you will be able to:

- Load and join multiple source files into a tidy analysis table.

- Clean messy columns (e.g., convert string prices like “$145” to numeric).

- Produce core EDA outputs: missingness, distributions, group summaries, and correlations.

- Compare price by room type and identify outliers using IQR.

- Analyze review recency (e.g., count listings with reviews by month).

- Use an AI coding assistant to draft/modify code—and then verify and interpret results yourself.

# **Important Note: The AI-generated solution may vary for each learner.**

#**Scenario:**
As a consultant working for a real estate start-up, you have collected Airbnb listing data from various sources to investigate the short-term rental market in New York. You'll analyze this data to provide insights on private rooms to the real estate company.

**Project Description**
New York City has a variety of Airbnb listings to meet the high demand for temporary lodging for travelers, with several different price levels, room types, and locations.

Practice your skills in importing, cleaning dataand report insights to a real estate start-up using AI features.



---



New York City is one of the most-visited cities in the world. There are many Airbnb listings in New York City to meet the high demand for temporary lodging for travelers, which can be anywhere between a few nights to many months. In this project, we will take a closer look at the New York Airbnb market by combining data from multiple file types like `.csv`, `.tsv`, and `.xlsx`.

**Dataset information:**

Three files containing data on 2019 Airbnb listings are available to you:

**data/airbnb_price.csv**
This is a CSV file containing data on Airbnb listing prices and locations.
- **`listing_id`**: unique identifier of listing
- **`price`**: nightly listing price in USD
- **`nbhood_full`**: name of borough and neighborhood where listing is located

**data/airbnb_room_type.xlsx**
This is an Excel file containing data on Airbnb listing descriptions and room types.
- **`listing_id`**: unique identifier of listing
- **`description`**: listing description
- **`room_type`**: Airbnb has three types of rooms: shared rooms, private rooms, and entire homes/apartments

**data/airbnb_last_review.tsv**
This is a TSV file containing data on Airbnb host names and review dates.
- **`listing_id`**: unique identifier of listing
- **`host_name`**: name of listing host
- **`last_review`**: date when the listing was last reviewed

In [1]:
# Import necessary packages such as pandas and numpy
import pandas as pd
import numpy as np

In [2]:
# Import CSV for prices
airbnb_price = pd.read_csv('airbnb_price.csv') #

# Import Excel file for room types
airbnb_room_type = pd.read_excel('airbnb_room_type.xlsx')

# Import TSV for review dates
airbnb_last_review = pd.read_csv('airbnb_last_review.tsv', sep='\t')

##**Instruction:**
Use following propmts for Exploring and analyzing the Airbnb data

**Example propmt 1:**

```
Task: Analyze the three Airbnb datasets (airbnb_last_review.tsv, airbnb_room_type.xlsx, and airbnb_price.csv).
Context: I want to understand whether these datasets can be related using a common column such as listing_id or similar identifier.
Expectation:

Identify potential common key columns.
Explain whether a relationship exists.
Provide Python pandas code to verify the relationship.
```



---
For Example Prompt 1, here's a complete response you could use.

Analysis

The three Airbnb datasets appear to describe different aspects of the same Airbnb listings:

airbnb_last_review.tsv contains information about the most recent review for each listing.
airbnb_room_type.xlsx contains the room type for each listing (for example, Entire home/apt, Private room, or Shared room).
airbnb_price.csv contains the listing price.

The most likely common key is listing_id, which uniquely identifies each Airbnb listing. If each dataset contains this column, they can be related using it.

To verify the relationship, inspect the column names and determine whether listing_id exists in all three datasets.

Python Code
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Display column names
print("Last Review Columns:")
print(last_review.columns)

print("\nRoom Type Columns:")
print(room_type.columns)

print("\nPrice Columns:")
print(price.columns)

# Check if listing_id exists in each dataset
print("\nDoes each dataset contain 'listing_id'?")

print("Last Review:", "listing_id" in last_review.columns)
print("Room Type:", "listing_id" in room_type.columns)
print("Price:", "listing_id" in price.columns)
Expected Conclusion

If listing_id appears in all three datasets:

The datasets can be related using listing_id.
This column serves as the primary key for merging the datasets.
A merge can then be performed using pd.merge() to combine the review, room type, and pricing information into a single DataFrame.


In [3]:
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Display column names
print("Last Review Columns:")
print(last_review.columns)

print("\nRoom Type Columns:")
print(room_type.columns)

print("\nPrice Columns:")
print(price.columns)

# Check if listing_id exists in each dataset
print("\nDoes each dataset contain 'listing_id'?")

print("Last Review:", "listing_id" in last_review.columns)
print("Room Type:", "listing_id" in room_type.columns)
print("Price:", "listing_id" in price.columns)

Last Review Columns:
Index(['listing_id', 'host_name', 'last_review'], dtype='object')

Room Type Columns:
Index(['listing_id', 'description', 'room_type'], dtype='object')

Price Columns:
Index(['listing_id', 'price', 'nbhood_full'], dtype='object')

Does each dataset contain 'listing_id'?
Last Review: True
Room Type: True
Price: True


**Example prompt 2:**


```
Task: Using pandas, Merge the three datasets into one DataFrame using the appropriate join key(s). Display the first 5 rows. Explain each merge step clearly.
```
Assuming you verified in Example Prompt 1 that the common key is listing_id, here's a complete solution for Example Prompt 2.

Python Code
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Merge the first two datasets on listing_id
merged_df = pd.merge(last_review, room_type, on="listing_id", how="inner")

# Merge the result with the price dataset
merged_df = pd.merge(merged_df, price, on="listing_id", how="inner")

# Display the first 5 rows
print(merged_df.head())
Explanation of Each Merge Step
Step 1: Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")
Reads the three Airbnb datasets into separate Pandas DataFrames.
sep="\t" is used because the first file is tab-separated.
read_excel() is used because the room type data is stored in an Excel file.
Step 2: Merge the review and room type datasets
merged_df = pd.merge(last_review, room_type,
                     on="listing_id",
                     how="inner")

This joins the two DataFrames using the common column listing_id.

on="listing_id" specifies the join key.
how="inner" keeps only listings that exist in both datasets.
Step 3: Merge with the price dataset
merged_df = pd.merge(merged_df,
                     price,
                     on="listing_id",
                     how="inner")

This combines the price information with the already merged data.

Again, an inner join keeps only listings that appear in all three datasets.

Step 4: Display the first five rows
print(merged_df.head())

The head() method displays the first five rows so you can verify that the merge was successful.

Expected Result

The final DataFrame should contain information from all three datasets, such as:

listing_id
host_name (if present)
last_review
room_type
price

Each row represents one Airbnb listing with all available information combined into a single table.

Note: If you receive a KeyError: 'listing_id', first run:

print(last_review.columns)
print(room_type.columns)
print(price.columns)

to verify the actual column names. If one dataset uses a different name (for example, id instead of listing_id), update the on= parameter accordingly.



In [4]:
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Merge the first two datasets on listing_id
merged_df = pd.merge(last_review, room_type, on="listing_id", how="inner")

# Merge the result with the price dataset
merged_df = pd.merge(merged_df, price, on="listing_id", how="inner")

# Display the first 5 rows
print(merged_df.head())

   listing_id    host_name   last_review  \
0        2595     Jennifer   May 21 2019   
1        3831  LisaRoxanne  July 05 2019   
2        5099        Chris  June 22 2019   
3        5178     Shunichi  June 24 2019   
4        5238          Ben  June 09 2019   

                                 description        room_type        price  \
0                      Skylit Midtown Castle  Entire home/apt  225 dollars   
1            Cozy Entire Floor of Brownstone  Entire home/apt   89 dollars   
2  Large Cozy 1 BR Apartment In Midtown East  Entire home/apt  200 dollars   
3            Large Furnished Room Near B'way     private room   79 dollars   
4         Cute & Cozy Lower East Side 1 bdrm  Entire home/apt  150 dollars   

                 nbhood_full  
0         Manhattan, Midtown  
1     Brooklyn, Clinton Hill  
2     Manhattan, Murray Hill  
3  Manhattan, Hell's Kitchen  
4       Manhattan, Chinatown  


**Example prompt 3:**


```
Task: Display the column names for each dataset using pandas.
```
For this prompt, you simply need to load the datasets and print the column names.

import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Display column names
print("Columns in airbnb_last_review.tsv:")
print(last_review.columns.tolist())

print("\nColumns in airbnb_room_type.xlsx:")
print(room_type.columns.tolist())

print("\nColumns in airbnb_price.csv:")
print(price.columns.tolist())
Explanation
columns returns the DataFrame's column names as a Pandas Index.
tolist() converts the Index into a regular Python list, making it easier to read.
Expected Output (example)
Columns in airbnb_last_review.tsv:
['listing_id', 'host_name', 'last_review']

Columns in airbnb_room_type.xlsx:
['listing_id', 'description', 'room_type']

Columns in airbnb_price.csv:
['listing_id', 'price', 'borough']

Note: The exact column names in your output may differ depending on the contents of your files. This code will display the actual column names in each dataset.

In [5]:
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Display column names
print("Columns in airbnb_last_review.tsv:")
print(last_review.columns.tolist())

print("\nColumns in airbnb_room_type.xlsx:")
print(room_type.columns.tolist())

print("\nColumns in airbnb_price.csv:")
print(price.columns.tolist())

Columns in airbnb_last_review.tsv:
['listing_id', 'host_name', 'last_review']

Columns in airbnb_room_type.xlsx:
['listing_id', 'description', 'room_type']

Columns in airbnb_price.csv:
['listing_id', 'price', 'nbhood_full']


**Example propmt 4:**


```
Task: After merging the three Airbnb datasets using listing_id, calculate how many records were lost during each join step.

Display:

The number of rows in each original dataset
The number of rows after each merge
The total number of records lost
```


---

Since this task asks you to calculate the number of records lost, you'll need to run the code and use your actual results. Here's the code:

import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Number of rows in each original dataset
print("Original dataset sizes:")
print("Last Review:", len(last_review))
print("Room Type:", len(room_type))
print("Price:", len(price))

# First merge
merge1 = pd.merge(last_review, room_type, on="listing_id", how="inner")
print("\nRows after first merge:", len(merge1))

# Second merge
merge2 = pd.merge(merge1, price, on="listing_id", how="inner")
print("Rows after second merge:", len(merge2))

# Records lost
lost_first = len(last_review) - len(merge1)
lost_second = len(merge1) - len(merge2)
total_lost = len(last_review) - len(merge2)

print("\nRecords lost after first merge:", lost_first)
print("Records lost after second merge:", lost_second)
print("Total records lost:", total_lost)
Explanation
len(dataframe) returns the number of rows in a DataFrame.
The first merge combines last_review and room_type using listing_id.
The second merge adds the price data.
The differences in row counts tell you how many records were excluded because they didn't have matching listing_id values in all datasets.



In [6]:
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Number of rows in each original dataset
print("Original dataset sizes:")
print("Last Review:", len(last_review))
print("Room Type:", len(room_type))
print("Price:", len(price))

# First merge
merge1 = pd.merge(last_review, room_type, on="listing_id", how="inner")
print("\nRows after first merge:", len(merge1))

# Second merge
merge2 = pd.merge(merge1, price, on="listing_id", how="inner")
print("Rows after second merge:", len(merge2))

# Records lost
lost_first = len(last_review) - len(merge1)
lost_second = len(merge1) - len(merge2)
total_lost = len(last_review) - len(merge2)

print("\nRecords lost after first merge:", lost_first)
print("Records lost after second merge:", lost_second)
print("Total records lost:", total_lost)

Original dataset sizes:
Last Review: 25209
Room Type: 25209
Price: 25209

Rows after first merge: 25209
Rows after second merge: 25209

Records lost after first merge: 0
Records lost after second merge: 0
Total records lost: 0


**Example prompt 5:**


```
Task: Using the airbnb_last_review.tsv dataset, write Python pandas code to:

Convert the last_review column to datetime format
Calculates the minimum and maximum review dates
Prints the results
```


---

import pandas as pd

# Load the dataset
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")

# Convert the last_review column to datetime format
last_review["last_review"] = pd.to_datetime(last_review["last_review"])

# Calculate the earliest and latest review dates
min_date = last_review["last_review"].min()
max_date = last_review["last_review"].max()

# Print the results
print("Earliest Review Date:", min_date)
print("Latest Review Date:", max_date)
Explanation
pd.to_datetime() converts the last_review column from text to a datetime format so that dates can be analyzed.
.min() finds the earliest review date.
.max() finds the most recent review date.
print() displays the results.
Expected Output

After running the code, you'll see something similar to:

Earliest Review Date: YYYY-MM-DD
Latest Review Date: YYYY-MM-DD






In [7]:
import pandas as pd

# Load the dataset
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")

# Convert the last_review column to datetime format
last_review["last_review"] = pd.to_datetime(last_review["last_review"])

# Calculate the earliest and latest review dates
min_date = last_review["last_review"].min()
max_date = last_review["last_review"].max()

# Print the results
print("Earliest Review Date:", min_date)
print("Latest Review Date:", max_date)

Earliest Review Date: 2019-01-01 00:00:00
Latest Review Date: 2019-07-09 00:00:00


**Example prompt 6:**



```
Task: After merging the three Airbnb datasets using listing_id, write Python pandas code to:

Standardize the room_type column to consistent capitalization
Count how many listings are classified as "Private Room"
Print the total count
```
Here's a clean solution for this prompt.

Python Code
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Merge the datasets
merged_df = pd.merge(last_review, room_type, on="listing_id", how="inner")
merged_df = pd.merge(merged_df, price, on="listing_id", how="inner")

# Standardize the room_type column
merged_df["room_type"] = merged_df["room_type"].str.title()

# Count the number of Private Room listings
private_room_count = (merged_df["room_type"] == "Private Room").sum()

# Print the result
print("Total Private Room listings:", private_room_count)
Explanation
str.title() standardizes the capitalization of the room_type values (for example, "private room" becomes "Private Room").
merged_df["room_type"] == "Private Room" creates a Boolean Series where each "Private Room" entry is True.
.sum() counts the number of True values, giving the total number of private room listings.


In [8]:
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Merge the datasets
merged_df = pd.merge(last_review, room_type, on="listing_id", how="inner")
merged_df = pd.merge(merged_df, price, on="listing_id", how="inner")

# Standardize the room_type column
merged_df["room_type"] = merged_df["room_type"].str.title()

# Count the number of Private Room listings
private_room_count = (merged_df["room_type"] == "Private Room").sum()

# Print the result
print("Total Private Room listings:", private_room_count)

Total Private Room listings: 11356


**Example prompt 7:**



```
Task: After merging the three Airbnb datasets using listing_id, write Python pandas code to: Calculate the average listing price Print the result clearly
```



---

import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Merge the datasets
merged_df = pd.merge(last_review, room_type, on="listing_id", how="inner")
merged_df = pd.merge(merged_df, price, on="listing_id", how="inner")

# Convert the price column to numeric
merged_df["price"] = (
    merged_df["price"]
    .replace(r"[\$,]", "", regex=True)
    .astype(float)
)

# Calculate the average listing price
average_price = merged_df["price"].mean()

# Print the result
print(f"Average Listing Price: ${average_price:.2f}")
Explanation
The three datasets are merged using the common key listing_id.
The price column is cleaned by removing the dollar sign ($) and any commas using .replace().
.astype(float) converts the cleaned prices into numeric values.
.mean() calculates the average listing price.
The result is printed with two decimal places using an f-string.

In [10]:
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Merge the datasets
merged_df = pd.merge(last_review, room_type, on="listing_id", how="inner")
merged_df = pd.merge(merged_df, price, on="listing_id", how="inner")

# Clean the price column
merged_df["price"] = (
    merged_df["price"]
    .str.replace(" dollars", "", regex=False)
    .str.replace(",", "", regex=False)
    .astype(float)
)

# Calculate the average listing price
average_price = merged_df["price"].mean()

# Print the result
print(f"Average Listing Price: ${average_price:.2f}")

Average Listing Price: $141.78


**Prompts for Descriptive pricing analysis**


- What’s the overall price distribution (median, IQR, tails)?

import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Merge the datasets
merged_df = pd.merge(last_review, room_type, on="listing_id")
merged_df = pd.merge(merged_df, price, on="listing_id")

# Clean the price column
merged_df["price"] = (
    merged_df["price"]
    .str.replace(" dollars", "", regex=False)
    .astype(float)
)

# Calculate statistics
minimum = merged_df["price"].min()
q1 = merged_df["price"].quantile(0.25)
median = merged_df["price"].median()
q3 = merged_df["price"].quantile(0.75)
maximum = merged_df["price"].max()
iqr = q3 - q1

# Print the results
print(f"Minimum Price: ${minimum:.2f}")
print(f"Q1 (25th percentile): ${q1:.2f}")
print(f"Median Price: ${median:.2f}")
print(f"Q3 (75th percentile): ${q3:.2f}")
print(f"IQR: ${iqr:.2f}")
print(f"Maximum Price: ${maximum:.2f}")

- Which room types are most/least expensive?

To determine which room types are the most and least expensive, calculate the average price for each room type.

Python Code
# Calculate the average price by room type
avg_price_by_room = (
    merged_df.groupby("room_type")["price"]
    .mean()
    .sort_values(ascending=False)
)

print("Average Price by Room Type:")
print(avg_price_by_room)
Optional (Cleaner Output)
# Display average prices rounded to two decimal places
avg_price_by_room = (
    merged_df.groupby("room_type")["price"]
    .mean()
    .round(2)
    .sort_values(ascending=False)
)

print(avg_price_by_room)

- Are there obvious outliers to cap or investigate?

Yes. Based on the statistics you've already calculated, there are clear outliers worth investigating.

Interpretation

The price data contains several potential outliers. The minimum price of $0 is unusual and may represent promotional listings, missing values, or data-entry errors. On the opposite end, the maximum price of $7,500 is much higher than the typical listing price.

Using the IQR method:

Q1 = $69
Q3 = $175
IQR = $106

The upper outlier threshold is:

Upper Bound = Q3 + 1.5 × IQR = 175 + (1.5 × 106) = $334

Any listing priced above $334 would be considered a potential outlier.

The lower outlier threshold is:

Lower Bound = Q1 − 1.5 × IQR = 69 − (1.5 × 106) = -$90

Since prices cannot be negative, there are no lower outliers by the IQR rule, although $0 listings should still be investigated.

Optional: Count the Outliers

You can also quantify them with this code:

# Calculate IQR bounds
Q1 = merged_df["price"].quantile(0.25)
Q3 = merged_df["price"].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Find outliers
outliers = merged_df[
    (merged_df["price"] < lower_bound) |
    (merged_df["price"] > upper_bound)
]

print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)
print("Number of Outliers:", len(outliers))

This will tell you exactly how many listings exceed the IQR threshold.

Conclusion

The dataset contains obvious high-price outliers, with prices exceeding $334 and reaching as high as $7,500. These listings should be reviewed to determine whether they represent legitimate luxury accommodations or data anomalies. Likewise, listings priced at $0 should be investigated, as they may indicate missing or incorrect pricing information. Depending on the purpose of the analysis, these extreme values could be capped, removed, or analyzed separately to prevent them from skewing summary statistics.

In [14]:
# Calculate IQR bounds
Q1 = merged_df["price"].quantile(0.25)
Q3 = merged_df["price"].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Find outliers
outliers = merged_df[
    (merged_df["price"] < lower_bound) |
    (merged_df["price"] > upper_bound)
]

print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)
print("Number of Outliers:", len(outliers))

Lower Bound: -90.0
Upper Bound: 334.0
Number of Outliers: 1326


In [13]:
# Standardize room type names
merged_df["room_type"] = merged_df["room_type"].str.title()

# Calculate average price by room type
avg_price_by_room = (
    merged_df.groupby("room_type")["price"]
    .mean()
    .round(2)
    .sort_values(ascending=False)
)

print(avg_price_by_room)

room_type
Entire Home/Apt    197.17
Private Room        81.64
Shared Room         53.47
Name: price, dtype: float64


In [11]:
import pandas as pd

# Load the datasets
last_review = pd.read_csv("airbnb_last_review.tsv", sep="\t")
room_type = pd.read_excel("airbnb_room_type.xlsx")
price = pd.read_csv("airbnb_price.csv")

# Merge the datasets
merged_df = pd.merge(last_review, room_type, on="listing_id")
merged_df = pd.merge(merged_df, price, on="listing_id")

# Clean the price column
merged_df["price"] = (
    merged_df["price"]
    .str.replace(" dollars", "", regex=False)
    .astype(float)
)

# Calculate statistics
minimum = merged_df["price"].min()
q1 = merged_df["price"].quantile(0.25)
median = merged_df["price"].median()
q3 = merged_df["price"].quantile(0.75)
maximum = merged_df["price"].max()
iqr = q3 - q1

# Print the results
print(f"Minimum Price: ${minimum:.2f}")
print(f"Q1 (25th percentile): ${q1:.2f}")
print(f"Median Price: ${median:.2f}")
print(f"Q3 (75th percentile): ${q3:.2f}")
print(f"IQR: ${iqr:.2f}")
print(f"Maximum Price: ${maximum:.2f}")

Minimum Price: $0.00
Q1 (25th percentile): $69.00
Median Price: $105.00
Q3 (75th percentile): $175.00
IQR: $106.00
Maximum Price: $7500.00


**Propmts for Inventory mix & positioning**


- What share of listings are Entire home/apt vs Private room vs others?

You can calculate the share of each room type with this code:

# Count and percentage of each room type
room_type_counts = merged_df["room_type"].value_counts()
room_type_percent = (
    merged_df["room_type"]
    .value_counts(normalize=True) * 100
).round(2)

room_type_summary = pd.DataFrame({
    "Count": room_type_counts,
    "Percentage": room_type_percent
})

print(room_type_summary)

 - Do some room types dominate certain price bands?

 To answer this question, you'll want to create price bands and see how room types are distributed within each band.

Use this code:

# Create price bands
price_bins = [0, 50, 100, 150, 200, 300, float('inf')]
price_labels = ["$0-50", "$51-100", "$101-150", "$151-200", "$201-300", "$301+"]

merged_df["price_band"] = pd.cut(
    merged_df["price"],
    bins=price_bins,
    labels=price_labels,
    include_lowest=True
)

# Count room types within each price band
room_price_dist = pd.crosstab(
    merged_df["price_band"],
    merged_df["room_type"],
    normalize="index"
) * 100

print(room_price_dist.round(2))

In [16]:
# Create price bands
price_bins = [0, 50, 100, 150, 200, 300, float('inf')]
price_labels = ["$0-50", "$51-100", "$101-150", "$151-200", "$201-300", "$301+"]

merged_df["price_band"] = pd.cut(
    merged_df["price"],
    bins=price_bins,
    labels=price_labels,
    include_lowest=True
)

# Count room types within each price band
room_price_dist = pd.crosstab(
    merged_df["price_band"],
    merged_df["room_type"],
    normalize="index"
) * 100

print(room_price_dist.round(2))

room_type   Entire Home/Apt  Private Room  Shared Room
price_band                                            
$0-50                  2.86         85.77        11.37
$51-100               25.57         72.42         2.00
$101-150              74.43         25.14         0.43
$151-200              90.53          9.47         0.00
$201-300              94.17          5.63         0.20
$301+                 94.03          5.84         0.13


In [15]:
# Count and percentage of each room type
room_type_counts = merged_df["room_type"].value_counts()
room_type_percent = (
    merged_df["room_type"]
    .value_counts(normalize=True) * 100
).round(2)

room_type_summary = pd.DataFrame({
    "Count": room_type_counts,
    "Percentage": room_type_percent
})

print(room_type_summary)

                 Count  Percentage
room_type                         
Entire Home/Apt  13266       52.62
Private Room     11356       45.05
Shared Room        587        2.33


**Propmt for Review recency & activity**

- How recent are reviews? Is demand cooling/heating?

Based on your earlier analysis, the review dates range from:

Earliest review: January 1, 2019
Latest review: July 9, 2019
Interpretation

The reviews in the dataset span approximately six months, from January to early July 2019. This indicates that the listings had relatively recent guest activity during the period covered by the data.

However, this dataset cannot determine whether demand is heating up or cooling down because it only contains each listing's most recent review date, not the full history or the number of reviews over time. Without monthly review counts or booking trends, there is not enough information to identify changes in demand.

Conclusion

The review data suggests that many listings were active during the first half of 2019, with the most recent review occurring on July 9, 2019. While the reviews are relatively recent within the dataset, there is insufficient evidence to conclude whether Airbnb demand was increasing or decreasing, since only the latest review for each listing is available rather than a complete time series of review activity.

- Are there seasonal peaks in last reviews?

To determine whether there are seasonal peaks, you'll need to count how many last reviews occurred in each month.

Run this code:

# Extract month name
merged_df["review_month"] = merged_df["last_review"].dt.month_name()

# Count reviews by month
monthly_reviews = (
    merged_df["review_month"]
    .value_counts()
    .reindex([
        "January", "February", "March",
        "April", "May", "June", "July"
    ])
)

print(monthly_reviews)

Or, if you'd like percentages:

monthly_review_pct = (
    merged_df["review_month"]
    .value_counts(normalize=True)
    .mul(100)
    .reindex([
        "January", "February", "March",
        "April", "May", "June", "July"
    ])
    .round(2)
)

print(monthly_review_pct)

In [19]:
# Convert last_review to datetime
merged_df["last_review"] = pd.to_datetime(merged_df["last_review"])

# Extract the month
merged_df["review_month"] = merged_df["last_review"].dt.month_name()

# Count reviews by month
monthly_reviews = (
    merged_df["review_month"]
    .value_counts()
    .reindex([
        "January", "February", "March",
        "April", "May", "June", "July"
    ])
)

print(monthly_reviews)

review_month
January      1440
February      470
March         820
April        1456
May          3697
June        12601
July         4725
Name: count, dtype: int64


**Propmts for Market segmentation**

- Are there price differences across room-type segments?

Yes. Based on the analysis you've already completed, there are clear price differences across room-type segments.

Average Nightly Price by Room Type
Room Type	Average Price
Entire Home/Apt	$197.17
Private Room	$81.64
Shared Room	$53.47
Interpretation

The results show substantial price differences among room types. Entire Home/Apt listings have the highest average nightly price at $197.17, which is more than twice the average price of Private Rooms ($81.64) and nearly four times that of Shared Rooms ($53.47).

This pricing pattern is also supported by the price-band analysis. Private Rooms dominate the lower-priced segments (under $100), while Entire Home/Apt listings account for more than 90% of listings priced above $150. Shared Rooms are concentrated almost entirely in the lowest price ranges and make up only a small share of the market.

Conclusion

There are significant price differences across room-type segments. Guests pay a premium for the privacy and exclusive use offered by Entire Home/Apt listings, while Private Rooms provide a more affordable alternative. Shared Rooms are the least expensive option and primarily serve budget-conscious travelers. These findings indicate that room type is a strong predictor of listing price.

- If we have neighborhood/borough (not seen here), how does location shift price?

Actually, your dataset appears to include a location field called nbhood_full, so you can analyze how location affects price.

To calculate the average price by neighborhood:

# Average price by neighborhood
avg_price_by_neighborhood = (
    merged_df.groupby("nbhood_full")["price"]
    .mean()
    .sort_values(ascending=False)
    .round(2)
)

print(avg_price_by_neighborhood.head(10))
print(avg_price_by_neighborhood.tail(10))

If you want to compare by borough (e.g., Manhattan, Brooklyn), you can extract the borough from nbhood_full if it is formatted like "Brooklyn, Williamsburg":

# Extract borough
merged_df["borough"] = merged_df["nbhood_full"].str.split(",", expand=True)[0]

# Average price by borough
borough_prices = (
    merged_df.groupby("borough")["price"]
    .agg(["count", "mean", "median"])
    .round(2)
    .sort_values("mean", ascending=False)
)

print(borough_prices)
Interpretation (if borough data is available)

Location typically has a strong influence on Airbnb pricing. Listings in central or highly desirable areas generally command higher nightly rates due to greater demand, proximity to attractions, and accessibility. Neighborhoods farther from major tourist or business districts tend to have lower average prices. Comparing average and median prices by borough or neighborhood can help identify premium locations and determine how much location contributes to price differences beyond room type.

In [20]:
# Average price by neighborhood
avg_price_by_neighborhood = (
    merged_df.groupby("nbhood_full")["price"]
    .mean()
    .sort_values(ascending=False)
    .round(2)
)

print(avg_price_by_neighborhood.head(10))
print(avg_price_by_neighborhood.tail(10))

nbhood_full
Brooklyn, Sea Gate              805.00
Manhattan, Tribeca              396.70
Manhattan, Flatiron District    342.06
Manhattan, NoHo                 335.02
Manhattan, SoHo                 299.16
Queens, Neponsit                274.67
Manhattan, Midtown              272.21
Manhattan, West Village         258.76
Staten Island, Willowbrook      249.00
Manhattan, Murray Hill          240.37
Name: price, dtype: float64
nbhood_full
Bronx, Soundview                 54.70
Bronx, Bronxdale                 53.60
Staten Island, Concord           53.30
Bronx, Olinville                 53.00
Staten Island, Port Richmond     48.80
Staten Island, Bull's Head       48.00
Queens, Little Neck              45.00
Bronx, Hunts Point               42.75
Bronx, Mount Eden                42.50
Staten Island, New Dorp Beach    38.00
Name: price, dtype: float64


In [21]:
# Extract borough
merged_df["borough"] = merged_df["nbhood_full"].str.split(",", expand=True)[0]

# Average price by borough
borough_prices = (
    merged_df.groupby("borough")["price"]
    .agg(["count", "mean", "median"])
    .round(2)
    .sort_values("mean", ascending=False)
)

print(borough_prices)

               count    mean  median
borough                             
Manhattan      10322  184.00   149.0
Brooklyn       10466  121.90    95.0
Queens          3456   92.81    70.0
Staten Island    267   86.04    71.0
Bronx            698   79.13    65.0



##**Submission Instructions**
- Submit your completed lab using the Start Assignment button on the assignment page in Canvas.
- Your submission can be include:
  - if you are using notebook then, all tasks should be written and submitted in a single notebook file, for example: (**your_name_labname.ipynb**).
  - if you are using python script file, all tasks should be written and submitted in a single python script file for example: **(your_name_labname.py)**.
- Add appropriate comments and any additional instructions if required.
